app/models/spree/invoice.rb

```diff
@@ -0,0 +1,128 @@
+class Spree::Invoice < ActiveRecord::Base
+  belongs_to :invoiceable, polymorphic: true
+  validates :invoiceable, presence: true
+
+  before_save :set_date
+  before_save :set_number, if: :use_sequential_number?
+  after_save :increase_invoice_number, if: :use_sequential_number?
+
+  # Delegates
+
+  delegate :bill_address,
+           :ship_address,
+           :tax_address,
+           :line_items,
+           :display_item_total,
+           :all_adjustments,
+           :shipments,
+           :display_total,
+           :payments, to: :invoiceable
```

Comment:
I would prefer to only have the smallest set of shared attributes in the delegations. Everything else could easily be done in the view with asking the invoiceable directly. I.e. Reimbursment does not have payments. Also the Invoice model must not be aware of the inner implementation of the invoiceable.